src/asyncresults.jl

```diff
@@ -0,0 +1,256 @@
+asynchronous loop.
+
+The result returned will be the [`Result`](@ref) of the last query run (the only query if
+using parameters).
+Any errors produced by the queries will be thrown together in a `CompositeException`.
+"""
+function handle_result(async_result::AsyncResult; throw_error=true)
+    errors = []
+    result = nothing
+    for result_ptr in _consume(async_result.jl_conn)
+        try
+            result = handle_result(
+                Result(
+                    result_ptr,
+                    async_result.jl_conn;
+                    async_result.result_kwargs[]...
+                );
+                throw_error=throw_error,
+            )
+        catch err
+            push!(errors, err)
+        end
+    end
+
+    if throw_error && !isempty(errors)
+        throw(CompositeException(errors))
+    elseif result === nothing
+        error(LOGGER, "Async query did not return result")
+    else
+        return result
+    end
+end
+
+function _consume(jl_conn::Connection)
+    async_result = jl_conn.async_result
+    result_ptrs = Ptr{libpq_c.PGresult}[]
+    watcher = FDWatcher(socket(jl_conn), true, false)  # can wait for reads
+    try
+        while true
+            if async_result.should_cancel
+                debug(LOGGER, "Received cancel signal for connection $(jl_conn.conn)")
+                _cancel(jl_conn)
+            end
+            debug(LOGGER, "Waiting to read from connection $(jl_conn.conn)")
+            wait(watcher)
+            debug(LOGGER, "Consuming input from connection $(jl_conn.conn)")
+            success = libpq_c.PQconsumeInput(jl_conn.conn) == 1
+            !success && error(error_message(jl_conn))
+
+            while libpq_c.PQisBusy(jl_conn.conn) == 0
+                debug(LOGGER, "Checking the result from connection $(jl_conn.conn)")
+                result_ptr = libpq_c.PQgetResult(jl_conn.conn)
+                if result_ptr == C_NULL
+                    debug(LOGGER, "Finished reading from connection $(jl_conn.conn)")
+                    return result_ptrs
+                else
+                    result_num = length(result_ptrs) + 1
+                    debug(LOGGER,
+                        "Saving result $result_num from connection $(jl_conn.conn)"
+                    )
+                    push!(result_ptrs, result_ptr)
+                end
+            end
+        end
+    finally
+        close(watcher)
+    end
+end
+
+"""
+    cancel(async_result::AsyncResult)
+
+If this [`AsyncResult`](@ref) represents a currently-executing query, attempt to cancel it.
+"""
+function cancel(async_result::AsyncResult)
+    async_result.should_cancel = true
+    return
+end
+
+function _cancel(jl_conn::Connection)
+    cancel_ptr = libpq_c.PQgetCancel(jl_conn.conn)
+    try
+        # https://www.postgresql.org/docs/10/libpq-cancel.html#LIBPQ-PQCANCEL
+        errbuf_size = 256
+        errbuf = fill(0x0, errbuf_size)
```

Comment:
```suggestion
        errbuf = zeros(UInt8, errbuf_size)
```

though really just personal preference

Reply:
No I prefer that too, I just blanked